src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -113,39 +120,57 @@ private void SetEntry(CacheEntry entry)
             // Initialize the last access timestamp at the time the entry is added
             entry.LastAccessed = utcNow;
 
-            var added = false;
             CacheEntry priorEntry;
+            if (_entries.TryGetValue(entry.Key, out priorEntry))
+            {
+                priorEntry.SetExpired(EvictionReason.Replaced);
+            }
 
-            _entryLock.EnterWriteLock();
-            try
+            if (!entry.CheckExpired(utcNow))
             {
+                var entryAdded = false;
 
-                if (_entries.TryGetValue(entry.Key, out priorEntry))
+                if (priorEntry == null)
                 {
-                    _entries.Remove(entry.Key);
-                    priorEntry.SetExpired(EvictionReason.Replaced);
+                    // Try to add the new entry if no previous entries exist.
+                    entryAdded = _entries.TryAdd(entry.Key, entry);
+                }
+                else
+                {
+                    // Try to update with the new entry if a previous entries exist.
+                    entryAdded = _entries.TryUpdate(entry.Key, entry, priorEntry);
+
+                    if (!entryAdded)
+                    {
+                        // The update will fail if the previous entry was removed after retrival.
+                        // Adding the new entry will succeed only if no entry has been added since.
+                        // This guarantees removing an old entry does not prevent adding a new entry.
+                        entryAdded = _entries.TryAdd(entry.Key, entry);
+                    }
                 }
 
-                if (!entry.CheckExpired(utcNow))
+                if (entryAdded)
```

Comment:
Spoke offline, There's no race here. We don't need to attach expiration tokens if the entry has already been removed and eviction callbacks are invoked.